File: tests/visual/a11y_check.mjs
// axe-core sweep for the docs pages — the WCAG checks that need a live DOM.
//
//   node tests/visual/a11y_check.mjs                # system Chrome
//   BOOST_CHROME_BIN=/path/to/chrome node ...       # explicit binary
//
// On macOS, point BOOST_CHROME_BIN at a `chrome-headless-shell` rather than at
// Google Chrome: Chrome's ProcessSingleton bind()s a unix socket and aborts "to
// avoid profile corruption" wherever that syscall is denied, and no flag turns
// it off. The shell binary has no ProcessSingleton.
//
// scripts/a11y_check.py is the always-on gate: it covers everything decidable
// from the markup (lang, alt, accessible names, duplicate ids, heading order)
// plus the 1.4.3 contrast ratios computed from the Aurora tokens, with no
// browser at all. This is the other half — the rules that only exist once the
// page is laid out and the accessibility tree is computed:
//
//   * color-contrast as RENDERED, including text over the glass panels and the
//     gradient, which no token-pair table can predict;
//   * ARIA attribute/role validity against the computed tree;
//   * landmark and region structure;
//   * elements hidden from assistive tech but still focusable.
//
// Runs against the LOCAL working tree (file://) like visual_check.mjs, so a PR's
// own HTML/CSS is what gets audited — no deploy needed.
import { launch } from "puppeteer-core";
import { readFileSync, existsSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "../..");

// Every page, not just the hand-authored four: the generated boards carry the
// same chrome and are where a heading-order or landmark slip actually shipped.
const PAGES = [
  "docs/index.html", "docs/adapters.html", "docs/eval.html", "docs/mcp-hub.html",
  "docs/roadmap.html", "docs/design-roadmap.html", "docs/commands.html",
  "docs/demo.html", "docs/chat.html", "docs/langchain.html",
];

// WCAG 2.1 A + AA. Deliberately not "best-practice": those are opinions, and a
// gate that fails on an opinion is a gate people learn to route around.
const TAGS = ["wcag2a", "wcag2aa", "wcag21a", "wcag21aa"];

const axePath = resolve(root, "tests/visual/node_modules/axe-core/axe.min.js");
if (!existsSync(axePath)) {
  console.error("a11y-check: axe-core not installed (npm install in tests/visual)");
  process.exit(2);
}
const axeSource = readFileSync(axePath, "utf8");

const CANDIDATE_BINS = [
  process.env.BOOST_CHROME_BIN,
  "/usr/bin/google-chrome",                                   // CI (ubuntu)
  "/usr/bin/chromium-browser",
  "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
].filter(Boolean);
const bin = CANDIDATE_BINS.find((p) => existsSync(p));
if (!bin) {
  console.error("a11y-check: no Chrome binary found (set BOOST_CHROME_BIN)");
  process.exit(2);
}

// chrome-headless-shell is the only binary that starts under a macOS sandbox
// that denies Mach port rendezvous — and it needs --single-process to do it, or
// it dies in bootstrap_check_in before the CDP pipe is ever ready and puppeteer
// reports a bare ProtocolError from launch(). visual_check.mjs already passes
// the flag, which is why that half of this directory ran locally and this half
// did not. Scoped to the shell binary on purpose rather than passed
// unconditionally: CI drives /usr/bin/google-chrome in NEW headless (this
// script takes puppeteer's default; visual_check.mjs asks for "shell"), and
// --single-process is a debug-only flag there whose interaction with new
// headless nothing in this repo exercises. Conditioning on the binary keeps
// the CI invocation byte-identical while making the local run work.
const needsSingleProcess = /chrome[-_]headless[-_]shell/.test(bin);

const browser = await launch({
  executablePath: bin,
  pipe: true,                       // no listening socket (sandbox-safe)
  // file:// pages load ../style/boost.css as a subresource; without this Chrome
  // treats each file as its own opaque origin and the stylesheet never applies,
  // which would make every contrast result meaningless.
  args: [
    "--allow-file-access-from-files", "--no-sandbox",
    ...(needsSingleProcess ? ["--single-process"] : []),
  ],
});

// ---------------------------------------------------------------------------
// Gradient contrast — the half axe cannot do.
//
// axe composites a solid background and gives up on anything else: text over a
// `linear-gradient` comes back as `incomplete`, never as a `violation`. This
// script fails on `result.violations` alone, so those nodes counted as passing
// — 153 of them across the docs, including `.cta`, `.grad`, `.btn-grad` and the
// hero stat numbers, which is to say every signature element of the Aurora
// system. The header CTA shipped at 1.37:1 (needs 4.5) and both gates were
// green; scripts/a11y_check.py missed it too, because it grades solid token
// pairs and a gradient is not one.
//
// A gradient has no single background colour, so grade against EVERY stop and
// keep the worst: text that passes over the light end of a gradient and fails
// over the dark end fails, because a reader gets whichever end their glyph
// happens to sit on.
//
// Two shapes, and they invert:
//   * text over a gradient panel  — grade the text colour against each stop;
//   * `background-clip: text`     — the gradient IS the text, so grade each
//                                   stop against whatever is painted behind it.
// Getting that backwards passes gradient text by comparing it to itself.
const GRADIENT_CONTRAST = () => {
  const parse = (s) => {
    const m = String(s).match(/rgba?\(([^)]+)\)/);
    if (!m) return null;
    const p = m[1].split(/[,\s/]+/).filter(Boolean).map(Number);
    if (p.length < 3 || p.some(Number.isNaN)) return null;
    return { r: p[0], g: p[1], b: p[2], a: p.length > 3 ? p[3] : 1 };
  };
  const lum = (c) => {
    const f = (v) => { v /= 255; return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4); };
    return 0.2126 * f(c.r) + 0.7152 * f(c.g) + 0.0722 * f(c.b);
  };
  const ratio = (a, b) => {
    const [x, y] = [lum(a), lum(b)];
    return (Math.max(x, y) + 0.05) / (Math.min(x, y) + 0.05);
  };
  const over = (fg, bg) => ({
    r: fg.r * fg.a + bg.r * (1 - fg.a),
    g: fg.g * fg.a + bg.g * (1 - fg.a),
    b: fg.b * fg.a + bg.b * (1 - fg.a), a: 1,
  });
  // The solid colour actually painted behind an element: the Aurora panels are
  // translucent, so the nearest ancestor with a background is rarely opaque and
  // taking it at face value would grade against a colour nobody sees.
  const CANVAS = { r: 7, g: 8, b: 15, a: 1 };          // --bg, the page ground
  const behind = (el) => {
    let acc = null;
    for (let n = el; n; n = n.parentElement) {
      const c = parse(getComputedStyle(n).backgroundColor);
      if (!c || c.a === 0) continue;
      acc = acc ? over(acc, c) : c;
      if (c.a === 1) break;
    }
    if (!acc) return CANVAS;
    return acc.a < 1 ? over(acc, CANVAS) : acc;
  };
  // Only elements holding their own text: a wrapper's colour says nothing about
  // a child that sets its own, and counting it double-reports every nesting.
  const hasOwnText = (el) =>
    [...el.childNodes].some((n) => n.nodeType === 3 && n.textContent.trim().length);

  const out = [];
  document.querySelectorAll("*").forEach((el) => {
    const cs = getComputedStyle(el);
    if (!/gradient/.test(cs.backgroundImage || "")) return;
    if (cs.visibility === "hidden" || cs.display === "none") return;
    const box = el.getBoundingClientRect();
    if (box.width < 1 || box.height < 1) return;
    if (!hasOwnText(el)) return;
    // Fully transparent stops are spacers in a fade, not colours a glyph sits on.
    const stops = (cs.backgroundImage.match(/rgba?\([^)]*\)/g) || [])
      .map(parse).filter((c) => c && c.a > 0.05);
    if (!stops.length) return;

    const px = parseFloat(cs.fontSize) || 16;
    const bold = (parseInt(cs.fontWeight, 10) || 400) >= 700;
    const large = px >= 24 || (px >= 18.66 && bold);   // WCAG 1.4.3 "large text"
    const need = large ? 3.0 : 4.5;
    const sel = el.tagName.toLowerCase() +
      (typeof el.className === "string" && el.className.trim()
        ? "." + el.className.trim().split(/\s+/).join(".") : "");
    const text = (el.textContent || "").trim().slice(0, 40);

    let got, kind;
    if ((cs.webkitBackgroundClip || cs.backgroundClip) === "text") {
      const bg = behind(el.parentElement || el);
      got = Math.min(...stops.map((s) => ratio(over(s, bg), bg)));
      kind = "gradient-text";
    } else {
      const fg = parse(cs.color);
      if (!fg) return;
      got = Math.min(...stops.map((s) => {
        const bg = over(s, behind(el));
        return ratio(over(fg, bg), bg);
      }));
      kind = "text-on-gradient";
    }
    out.push({ kind, sel, text, need, got: Math.round(got * 100) / 100 });
  });
  return out;
};

let violatingNodes = 0;
let gradientsChecked = 0;
for (const rel of PAGES) {
  const file = resolve(root, rel);
  if (!existsSync(file)) {
    console.error(`a11y-check: ${rel} is missing`);
    violatingNodes++;
    continue;
  }
  const page = await browser.newPage();
  await page.setViewport({ width: 1280, height: 900 });
  // Emulate prefers-reduced-motion BEFORE navigating. `.js .reveal` fades in
  // over 600ms (opacity 0 -> 1), and axe computes contrast from the COMPOSITED
  // colour — so running mid-transition reports --text-3 (#767c96, a genuine
  // 4.85:1) as whatever partial-opacity blend it caught: #363948, #505467 and
  // #656a81 were all observed, at 20%, 44% and 72% of the token's luminance.
  // None of those exist anywhere in the repo, which is what made the failures
  // look like a phantom.
  //
  // boost.css already handles this: its `prefers-reduced-motion: reduce` block
  // forces `opacity: 1 !important` on .reveal. The sweep simply never asked for
  // it. Emulating the feature is also more honest than waiting out the
  // animation — it audits the page as a motion-sensitive user receives it.
  await page.emulateMediaFeatures([
    { name: "prefers-reduced-motion", value: "reduce" },
  ]);
  await page.goto("file://" + file, { waitUntil: "load" });

  // Guard: prove the page is actually styled before trusting a contrast result.
  // Most pages link ../style/boost.css RELATIVELY over file://, which only
  // works because of --allow-file-access-from-files, and that flag has
  // intermittently not taken effect on the runner. Colours then fall back to
  // browser defaults and axe reports a dozen confident, specific, entirely
  // fictional violations against values that appear nowhere in this repo
  // (#505467, #656a81 — the real --text-3 is #767c96 at 4.85:1, comfortably AA).
  //
  // Probe --bg, NOT --text-3: docs/commands.html is generated self-contained
  // with its tokens inline and defines no --text-3 at all, so keying on that
  // token failed a page that was perfectly styled. --bg is defined by both the
  // shared sheet and every inline :root block, so it detects an unstyled page
  // without assuming where the styling came from.
  //
  // Failing here is the whole point: "not styled" is one line a maintainer can
  // act on, where the alternative sends them auditing a colour that has been
  // correct for months.
  const styled = await page.evaluate(() =>
    getComputedStyle(document.documentElement)
      .getPropertyValue("--bg").trim());
  if (!styled) {
    console.error(`FAIL ${rel} — page is unstyled (--bg unresolved); `
      + `every contrast result here would be meaningless, so it is reported `
      + `as a load failure rather than as violations`);
    violatingNodes++;
    await page.close();
    continue;
  }

  await page.evaluate(axeSource);
  const result = await page.evaluate(
    async (tags) => await window.axe.run(document, { runOnly: { type: "tag", values: tags } }),
    TAGS,
  );

  const gradients = await page.evaluate(GRADIENT_CONTRAST);
  gradientsChecked += gradients.length;
  const dim = gradients.filter((g) => g.got < g.need);
  if (dim.length) {
    console.error(`FAIL ${rel} — ${dim.length} gradient element(s) below AA `
      + `(axe reports these as "incomplete", never as violations)`);
    for (const g of dim) {
      console.error(`      ${g.got}:1 (needs ${g.need}) ${g.kind} ${g.sel} — "${g.text}"`);
    }
  }
  violatingNodes += dim.length;

  const violations = result.violations;
  const count = violations.reduce((n, v) => n + v.nodes.length, 0);
  violatingNodes += count;
  if (count) {
    console.error(`FAIL ${rel} — ${count} violating node(s)`);
    for (const v of violations) {
      console.error(`  [${v.impact}] ${v.id}: ${v.help}`);
      for (const node of v.nodes.slice(0, 3)) {
        console.error(`      ${node.target.join(" ")}`);
        // failureSummary is axe's own "to fix this" text — the single most
        // useful line for whoever has to act on the failure.
        if (node.failureSummary) {
          console.error(`      ${node.failureSummary.replace(/\n\s*/g, " ")}`);
        }
      }
      if (v.nodes.length > 3) console.error(`      … and ${v.nodes.length - 3} more`);
    }
  } else if (!dim.length) {
    console.log(`PASS ${rel} — ${result.passes.length} rules passed`
      + `, ${gradients.length} gradient element(s) graded`);
  }
  await page.close();
}

await browser.close();
if (violatingNodes) {
  console.error(`a11y-check: ${violatingNodes} violating node(s) across ${PAGES.length} pages`);
  process.exit(1);
}
console.log(`a11y-check: OK — ${PAGES.length} pages clean against ${TAGS.join(", ")}`
  + `, plus ${gradientsChecked} gradient element(s) graded per colour stop`);

File: scripts/a11y_check.py
#!/usr/bin/env python3
# Copyright the boost contributors.
# SPDX-License-Identifier: Apache-2.0
"""Accessibility gate for the docs pages — WCAG 2.1 AA, no browser required.

    python3 scripts/a11y_check.py           # audit docs/*.html
    python3 scripts/a11y_check.py -v        # list every page checked
    python3 scripts/a11y_check.py --list    # list the rules and exit

Two halves, both pure stdlib so this runs in the lint job beside the other
`--check` gates rather than needing Node, Chrome or a network fetch:

* **Structure** — the WCAG failures that are decidable from the markup alone:
  a missing ``lang``, an image with no ``alt``, a link or button with no
  accessible name, a duplicate ``id``, a skipped heading level, a form control
  with nothing to label it.
* **Contrast** — the 1.4.3 ratios for the Aurora tokens, computed from
  ``style/boost.css`` itself, for the ink/ground pairs the CSS actually puts
  together. The palette leans on muted ``--text-2``/``--text-3`` against dark
  glass, which is exactly where AA slips.

What this deliberately does NOT cover is anything that needs layout or a live
DOM: focus order, visible focus rings, ARIA state as computed, reflow at 320px.
Those are the axe-core sweep's job (tests/visual/), which needs a real browser
and therefore cannot be a cheap always-on gate. Static analysis is the floor,
not the ceiling — see docs/DEBUGGING.md.
"""
from __future__ import annotations

import argparse
import re
from html.parser import HTMLParser
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DOCS = ROOT / "docs"
CSS = ROOT / "style" / "boost.css"

# WCAG 1.4.3: normal text needs 4.5:1, large text (>=24px, or >=18.66px bold)
# needs 3.0:1. Everything audited here is body-sized, so 4.5 is the floor.
AA_NORMAL = 4.5

RULES = {
    "html-has-lang": "<html> carries a lang attribute (WCAG 3.1.1)",
    "image-alt": "every <img> has an alt attribute (WCAG 1.1.1)",
    "link-name": "every <a> has an accessible name (WCAG 2.4.4)",
    "button-name": "every <button> has an accessible name (WCAG 4.1.2)",
    "form-label": "every form control can be labelled (WCAG 3.3.2)",
    "duplicate-id": "ids are unique (WCAG 4.1.1)",
    "heading-order": "heading levels never skip (WCAG 1.3.1)",
    "page-has-h1": "each page has exactly one <h1> (WCAG 1.3.1)",
    "contrast": "text/background pairs meet %.1f:1 (WCAG 1.4.3)" % AA_NORMAL,
}

# Void elements never have an end tag, so the parser must not wait for one.
VOID = {"area", "base", "br", "col", "embed", "hr", "img", "input", "link",
        "meta", "param", "source", "track", "wbr"}


class _Auditor(HTMLParser):
    """Collect structural WCAG findings from one page's markup."""

    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.findings: list = []
        self._ids: dict = {}
        self._headings: list = []
        self._link = None
        self._link_text = ""
        self._button = None
        self._button_text = ""

    # -- helpers ---------------------------------------------------------
    def _add(self, rule: str, message: str, line=None) -> None:
        self.findings.append((line or self.getpos()[0], rule, message))

    # -- tag handling ----------------------------------------------------
    def handle_starttag(self, tag, attrs) -> None:
        a = dict(attrs)
        line = self.getpos()[0]

        if tag == "html" and not (a.get("lang") or "").strip():
            self._add("html-has-lang", "<html> has no lang attribute", line)
        if tag == "img" and "alt" not in a:
            self._add("image-alt",
                      "<img src=%r> has no alt attribute" % a.get("src", "?"), line)
        if "id" in a:
            prior = self._ids.get(a["id"])
            if prior:
                self._add("duplicate-id",
                          "id=%r first used on line %d" % (a["id"], prior), line)
            else:
                self._ids[a["id"]] = line
        if tag in ("h1", "h2", "h3", "h4", "h5", "h6"):
            self._headings.append((line, int(tag[1])))
        if tag == "a":
            self._link = (line, a)
            self._link_text = ""
        if tag == "button":
            self._button = (line, a)
            self._button_text = ""
        if tag in ("input", "select", "textarea"):
            # A control is labellable via aria-*, a title, or an id a <label for>
            # can point at. Hidden and submit-style inputs carry their own name.
            kind = (a.get("type") or "").lower()
            if kind in ("hidden", "submit", "reset", "button", "image"):
                return
            if not any(a.get(k) for k in ("aria-label", "aria-labelledby",
                                          "id", "title")):
                self._add("form-label",
                          "<%s> has nothing to associate a label with" % tag, line)

    def handle_startendtag(self, tag, attrs) -> None:
        self.handle_starttag(tag, attrs)

    def handle_data(self, data) -> None:
        if self._link is not None:
            self._link_text += data
        if self._button is not None:
            self._button_text += data

    def handle_endtag(self, tag) -> None:
        if tag == "a" and self._link is not None:
            line, a = self._link
            if not self._accessible_name(self._link_text, a):
                self._add("link-name",
                          "<a href=%r> has no accessible name"
                          % a.get("href", "?"), line)
            self._link = None
        if tag == "button" and self._button is not None:
            line, a = self._button
            if not self._accessible_name(self._button_text, a):
                self._add("button-name", "<button> has no accessible name", line)
            self._button = None

    @staticmethod
    def _accessible_name(text: str, attrs: dict) -> str:
        """The name a screen reader would announce, or "" if there is none."""
        return (text.strip()
                or (attrs.get("aria-label") or "").strip()
                or (attrs.get("aria-labelledby") or "").strip()
                or (attrs.get("title") or "").strip())

    # -- whole-document checks -------------------------------------------
    def finish(self) -> list:
        """Findings that need the whole document, appended to the per-tag ones."""
        levels = [lvl for _line, lvl in self._headings]
        if levels.count(1) != 1:
            self._add("page-has-h1",
                      "page has %d <h1> elements, expected exactly 1"
                      % levels.count(1), 1)
        previous = 0
        for line, level in self._headings:
            if previous and level > previous + 1:
                self._add("heading-order",
                          "h%d follows h%d — skips h%d"
                          % (level, previous, previous + 1), line)
            previous = level
        return sorted(self.findings)


# --------------------------------------------------------------------------
# contrast
# --------------------------------------------------------------------------
_HEX = re.compile(r"^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$")


def _channel(value: int) -> float:
    """One sRGB channel, linearized per WCAG's relative-luminance formula."""
    c = value / 255
    return c / 12.92 if c <= 0.04045 else ((c + 0.055) / 1.055) ** 2.4


def luminance(hex_color: str) -> float:
    """Relative luminance of a #rgb / #rrggbb color (WCAG 2.1 definition)."""
    h = hex_color.lstrip("#")
    if len(h) == 3:
        h = "".join(ch * 2 for ch in h)
    r, g, b = (int(h[i:i + 2], 16) for i in (0, 2, 4))
    return 0.2126 * _channel(r) + 0.7152 * _channel(g) + 0.0722 * _channel(b)


def contrast(a: str, b: str) -> float:
    """Contrast ratio between two colors — 1.0 (identical) to 21.0 (b/w)."""
    la, lb = luminance(a), luminance(b)
    hi, lo = max(la, lb), min(la, lb)
    return (hi + 0.05) / (lo + 0.05)


def read_tokens(css_text: str) -> dict:
    """``{--token: #hex}`` for every plain hex custom property in the CSS."""
    tokens = {}
    for name, value in re.findall(r"(--[\w-]+)\s*:\s*([^;]+);", css_text):
        value = value.strip()
        if _HEX.match(value):
            tokens[name] = value
    return tokens


# The ink/ground pairs the stylesheet actually renders together. Listing them
# explicitly rather than testing the cross-product keeps the gate honest: a pair
# that never co-occurs is not a real contrast failure, and flagging it would
# train people to ignore this check.
#
# The hover surface is checked as well as the resting one. Only listing the
# resting surface let --text-3 pass at 4.53:1 and then fail at 4.15:1 the moment
# a card lit up under the cursor — the gate was blind to the state a reader is
# actually in while reading a card.
PAIRS = (
    ("--text", "--bg"), ("--text-2", "--bg"), ("--text-3", "--bg"),
    ("--text", "--surface-1"), ("--text-2", "--surface-1"),
    ("--text-3", "--surface-1"),
    ("--text", "--surface-2"), ("--text-2", "--surface-2"),
    ("--text-3", "--surface-2"),
    ("--text", "--surface-3"), ("--text-2", "--surface-3"),
    ("--text-3", "--surface-3"),
    ("--text", "--panel-solid"), ("--text-2", "--panel-solid"),
    ("--text-3", "--panel-solid"),
    ("--cyan", "--bg"), ("--violet", "--bg"), ("--pink", "--bg"),
    ("--blue", "--bg"), ("--sky", "--bg"), ("--green", "--bg"),
    ("--yellow", "--bg"),
    ("--text", "--term-bg"), ("--text-2", "--term-bg"),
)


def check_contrast(css_text: str) -> list:
    """Findings for every declared pair below the AA floor."""
    tokens = read_tokens(css_text)
    findings = []
    for ink, ground in PAIRS:
        if ink not in tokens or ground not in tokens:
            findings.append((0, "contrast",
                             "pair (%s, %s) names a token that no longer exists"
                             % (ink, ground)))
            continue
        ratio = contrast(tokens[ink], tokens[ground])
        if ratio < AA_NORMAL:
            findings.append((0, "contrast",
                             "%s (%s) on %s (%s) is %.2f:1, below %.1f:1"
                             % (ink, tokens[ink], ground, tokens[ground],
                                ratio, AA_NORMAL)))
    return findings


# --------------------------------------------------------------------------
def audit_page(path: Path) -> list:
    """Structural findings for one HTML file."""
    parser = _Auditor()
    parser.feed(path.read_text(encoding="utf-8", errors="replace"))
    return parser.finish()


def main(argv=None) -> int:
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("-v", "--verbose", action="store_true",
                    help="list every page checked, not just failures")
    ap.add_argument("--list", action="store_true",
                    help="list the rules this gate enforces and exit")
    args = ap.parse_args(argv)

    if args.list:
        for rule, description in sorted(RULES.items()):
            print("  %-16s %s" % (rule, description))
        return 0

    pages = sorted(DOCS.glob("*.html"))
    if not pages:
        print("a11y: no pages found under %s" % DOCS)
        return 2

    failures = 0
    for page in pages:
        findings = audit_page(page)
        if findings:
            failures += len(findings)
            print("%s — %d issue%s"
                  % (page.relative_to(ROOT), len(findings),
                     "" if len(findings) == 1 else "s"))
            for line, rule, message in findings:
                print("  L%-5d %-14s %s" % (line, rule, message))
        elif args.verbose:
            print("%s — clean" % page.relative_to(ROOT))

    contrast_findings = check_contrast(CSS.read_text(encoding="utf-8"))
    if contrast_findings:
        failures += len(contrast_findings)
        print("%s — %d contrast issue%s"
              % (CSS.relative_to(ROOT), len(contrast_findings),
                 "" if len(contrast_findings) == 1 else "s"))
        for _line, rule, message in contrast_findings:
            print("  %-14s %s" % (rule, message))
    elif args.verbose:
        print("%s — %d contrast pairs clean" % (CSS.relative_to(ROOT), len(PAIRS)))

    if failures:
        print("\na11y: FAIL — %d issue%s across %d page%s"
              % (failures, "" if failures == 1 else "s",
                 len(pages), "" if len(pages) == 1 else "s"))
        return 1
    print("a11y: OK — %d pages and %d contrast pairs meet WCAG 2.1 AA"
          % (len(pages), len(PAIRS)))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
